README.md
# CompletePaymentSystems

Ruby gem for a simple CPS (Complete Payment Systems) service API - creates request XML from a parameters hash, processes the response.

Only direct-3D payments supported for now.

###### Workflow Overview:
1. Install gem, congifure defaults
2. Build an appropriate parameter hash using user-entered card values
3. Instantiate CPS::Request object passing the hash as the single argument
4. Pass the request instance to the form view (see /example_request_form.html.erb)
5. Capture the CPS server response in a controller action that listens on _config.default_callback_url_
6. Based on the response (@response.ok?) change order status
7. Based on order status, process user return experience at _config.default_redirect_url_

## Installation

Add this line to your application's Gemfile:

    gem 'complete_payment_systems'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install complete_payment_systems

## Usage

The complete module name is CompletePaymentSystems, but the aliased shorthand CPS is recommended.

### Configure Defaults
in */config/initializers/cps.rb*

```ruby
CPS.configure do |config|
  # Your CPS account user
  config.default_user = "pasta_test_3d"

  # server-to-server POST response processing route
  config.default_callback_url = "http://www.app.com/cps/process"

  # Url to "Thank-you"-"Retry" differ route
  config.default_redirect_url = "http://www.app.com/cps/return"
  config.default_product_name = "Product"
  config.default_product_url = "www.test.com"
  config.default_country = "US"

  # Used in client data, like street and zip, if you choose not to ask for these
  config.placeholder_value = "Not@Used.com"

  config.cps_url = "https://3ds.cps.lv/GatorServo/request" # This will probably never change
  config.cps_method = "sendForAuth" # This will also probably remain the same
  config.cert_pass = 'pasS%123' # your cert chain password

  # Your .pem format cert chain location
  config.rsa_cert_path = "#{CPS.root}/lib/complete_payment_systems/certs/Pasta_test_3d.pem"
end
```

### Build a payment request
Instantiate a CPS::Request object, pass it a hash with parameters:
```ruby
hash = {
  order: (Time.now.to_i),            # Pass the unique purchase ID here
  value: 100     ,                   # Pass the purchase value in cents here (1$ purcase value = 100)
  currency: "EUR",                   # Pass the purchase currency 3-letter code here ($ = "USD")
  holder_name: "John",               # Ask buyer for this in a form
  holder_surname: "Doe",             # Ask buyer for this in a form
  card_number: "4314229999999913",   # Ask buyer for this in a form
  card_exp: "01/18",                 # Ask buyer for this in a form
  card_cvv: "123",                   # Ask buyer for this in a form
  holder_ip: "123.124.125.226",      # Get this from request.remote_ip
  holder_street: "NOT USED",         # (Optional) Ask buyer for this in a form
  holder_zip: "NOT USED",            # (Optional) Ask buyer for this in a form
  holder_city: "NOT USED",           # (Optional) Ask buyer for this in a form
  holder_country: "US",              # (Optional) Ask buyer for this in a form
  holder_email: "test@domain.com",   # (Optional) Ask buyer for this in a form
  user: "Some User"                  # Best defined in config and not passed
  callback_url: "http://www.app.lv", # Best defined in config and not passed
  redirect_url: "http://www.app.lv", # Best defined in config and not passed
  product_name: "Product",           # Best defined in config and not passed
  product_url: "www.test.com"        # Best defined in config and not passed
}
CPS::Request.new(hash)
```
# Process response
Instantiate a CPS::Response object with an xml parameter (get the xml from server response.body)
```ruby
@response = CPS.Response.new(request_response.body)
```
API exposes reader methods #xml, #response_hash, #code, #message
as well as convenience methods #ok? and #signature_ok?

```ruby
r = CPS::Response.new(valid_xml_response)
r.response_hash
  => {"referenceId"=>"9095359",
 "orderId"=>"1422019383",
 "value"=>"100",
 "currency"=>"USD",
 "resultCode"=>"000",
 "resultMessage"=>"Captured",
 "resultText"=>"Captured",
 "digiSignature"=>
  "HlcnASe5fZyM7uVz4xwqp ... XGx5SVqSA=="}
r.code
  => "000"
r.message
  => "Captured"
r.signature_ok?
  => true
r.ok?
  => true
```

## Contributing

1. Fork it ( https://github.com/[my-github-username]/complete_payment_systems/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## Dev Notes

calbackUrl atbilde atnāks pirmā, jo te mēs sūtam atbildi pa tiešo no mūsu servera uz jūso noradītu URL

#### Direct-3D request example
```ruby
hash = {
  order: (Time.now.to_i),            # Pass the unique purchase ID here
  value: 166,                        # Pass the purchase value in cents here (1$ purcase value = 100)
  currency: (params["currency"]),    # Pass the purchase currency 3-letter code here ($ = "USD")
  holder_name: "John",               # Ask buyer for this in a form
  holder_surname: "Doe",             # Ask buyer for this in a form
  card_number: "4314229999999913",   # Ask buyer for this in a form
  card_exp: "01/18",                 # Ask buyer for this in a form
  card_cvv: "123",                   # Ask buyer for this in a form
  holder_ip: "123.124.125.226",      # Get this from request.remote_ip
  holder_street: "NOT USED",         # (Optional) Ask buyer for this in a form
  holder_zip: "NOT USED",            # (Optional) Ask buyer for this in a form
  holder_city: "NOT USED",           # (Optional) Ask buyer for this in a form
  holder_country: "US",              # (Optional) Ask buyer for this in a form
  holder_email: "test@domain.com",   # (Optional) Ask buyer for this in a form
  user: "Some User"                  # Best defined in config and not passed
  callback_url: "http://www.app.lv", # Best defined in config and not passed
  redirect_url: "http://www.app.lv", # Best defined in config and not passed
  product_name: "Product",           # Best defined in config and not passed
  product_url: "www.test.com"        # Best defined in config and not passed
}
@request = CPS::Request.new(hash)
```

CHANGELOG.md
# v0.0.4

Cleaned up test case initialization

# v0.0.2

Added default values to named parameters for Ruby v2.0.0 compatability

# v0.0.1

First legitimate release

spec/spec_helper.rb
require 'bundler/setup'
Bundler.setup

require "complete_payment_systems"

RSpec.configure do |config|
  # some (optional) config here
end

spec/tests/foo_spec.rb
# rspec spec/tests/foo_spec.rb

describe "Gem_Name" do
  pending "write it"
  it "a" do
    expect(:a).to eq :a
  end
end

spec/tests/request_spec.rb
# rspec spec/tests/request_spec.rb

describe "Test request" do
  it "should be successful" do
    expect(CPS.unirest).to eq "OK"
  end

end

lib/complete_payment_systems/xml_parsing.rb
module CompletePaymentSystems

  def self.parse_response xml

    @response = Nokogiri::XML(xml) do |config|
      config.strict.nonet
    end

    puts @response.xpath("//referenceId").text
    puts @response.xpath("//orderId").text
    puts @response.xpath("//value").text
    puts @response.xpath("//currency").text
    puts @response.xpath("//resultCode").text
    puts @response.xpath("//resultMessage").text # Important
    puts @response.xpath("//resultText").text # Informative only
    puts @response.xpath("//digiSignature").text

    return {
      "referenceId" => @response.xpath("//referenceId").text,
      "orderId" => @response.xpath("//orderId").text,
      "value" => @response.xpath("//value").text,
      "currency" => @response.xpath("//currency").text,
      "resultCode" => @response.xpath("//resultCode").text,
      "resultMessage" => @response.xpath("//resultMessage").text,
      "resultText" => @response.xpath("//resultText").text,
      "digiSignature" => @response.xpath("//digiSignature").text
    }

  end

  # CPS.verify_signature(CPS.parse_response(CPS.read_test_xml)["digiSignature"], CPS.build_hashable_string( CPS.parse_response(CPS.read_test_xml) ) )
  def self.verify_signature(digi_signature, hash_string)
    puts "called verify signature"
    puts digi_signature
    puts hash_string

    puts "Decoded resp:"
    puts decoded_resp = Base64.decode64(digi_signature)

    puts "Public key:"
    puts public_key = OpenSSL::X509::Certificate.new(
      File.read "#{CPS.root}/lib/complete_payment_systems/certs/cps.cer").
      public_key

    puts "Verification status:"
    puts public_key.verify(OpenSSL::Digest::SHA1.new, decoded_resp, hash_string).inspect
  end

  # CPS.build_hashable_string( CPS.parse_response(CPS.read_test_xml) )
  def self.build_hashable_string response_hash
    rh = response_hash
    return "#{rh["referenceId"]}#{rh["orderId"]}#{rh["value"]}#{rh["currency"]}#{rh["resultCode"]}#{rh["resultMessage"]}#{rh["resultText"]}"
  end

  def self.read_test_xml
    return File.read("#{CPS.root}/response.xml")
  end

end

lib/complete_payment_systems/request_processing.rb
module CompletePaymentSystems
  class Request

    attr_reader :params, :xml

    def initialize(params)
      @params = params #|| File.read("#{CPS.root}/response.xml")
      @xml = build_request_xml
    end

    def send!
      response = Unirest.post CPS.config.cps_url,
                          #headers:{ "Accept" => "application/json" },
                          parameters:{ type: CPS.config.cps_method, xml: xml }

      puts response.code # Status code
      puts response.headers # Response headers
      puts response.body # Parsed body

      return "OK" if response.body.to_s.match(CPS.config.success_regex).present?
      return "ERROR"
    end

    private
      def sign(type: CPS.config.cps_method, user: CPS.config.default_user, order_id: "", value: "", currency: "", card_number: "", product: CPS.config.default_product_name)
        sign_string = [type, user, order_id, value, currency, card_number, product].join()
        rsa         = OpenSSL::PKey::RSA.new(File.read(CPS.config.rsa_cert_path), CPS.config.cert_pass )
        signed_hash = rsa.sign(OpenSSL::Digest::SHA1.new, sign_string)
        Base64.encode64(signed_hash)
      end

      def build_request_xml
        values = {
          user: (params[:user] || CPS.config.default_user), # "test_pasta_sign" for direct, "pasta_test_3d" for direct with 3D
          callback_url: (params[:callback_url] || CPS.config.default_callback_url),
          redirect_url: (params[:redirect_url] || CPS.config.default_redirect_url),
          order: (params[:order] || Time.now.to_i), # TO-DO remove fallback
          value: (params[:value]),
          currency: (params[:currency]),
          holder_name: params[:holder_name],
          holder_surname: params[:holder_surname],
          holder_street: (params[:holder_street] || CPS.config.placeholder_value),
          holder_zip: (params[:holder_zip] || CPS.config.placeholder_value),
          holder_city: (params[:holder_city] || CPS.config.placeholder_value),
          holder_country: (params[:holder_country] || CPS.config.default_country),
          holder_email: (params[:holder_email] || CPS.config.placeholder_value),
          holder_ip: ( params[:holder_ip] ),
          card_number: params[:card_number], # 4012001037167778 for direct, For 3D-Secure number is 4314229999999913
          card_exp: params[:card_exp], # 06/18 for direct, 3D-sec is "01/18"
          card_cvv: params[:card_cvv], # 999 for direct, 3D-sec is "123"
          product_name: (params[:product_name] || CPS.config.default_product_name),
          product_url: (params[:product_url] || CPS.config.default_product_url)
        }

        values[:signature] = sign(order_id: values[:order], value: values[:value], currency: values[:currency], card_number: values[:card_number]).gsub(/\n/, '')

        values[:signature_line] = '<digiSignature>' + values[:signature] + '</digiSignature>'

        xml_string = ""
        xml_string = %Q|<?xml version="1.0" encoding="UTF-8" ?>\n| +
          %Q|<cpsxml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"\n| +
          %Q|        xsi:schemaLocation="http://www.cps.lv/xml/ns/cpsxml https://3ds.cps.lv/GatorServo/Gator_SendForAuth.xsd"\n| +
          %Q|        xmlns="http://www.cps.lv/xml/ns/cpsxml">\n| +
          %Q|  <header xmlns="">\n| +
          %Q|    <responsetype>direct</responsetype>\n| +
          %Q|    <user>#{values[:user]}</user>\n| +
          %Q|    <type>sendForAuth</type>\n| +
          %Q|    <transType>DB</transType>\n| +
          %Q|    | + values[:signature_line] + "\n" +
          %Q|    <callbackUrl>#{values[:callback_url]}</callbackUrl>\n| +
          %Q|    <redirectUrl>#{values[:redirect_url]}</redirectUrl>\n| +
          %Q|  </header>\n| +
          %Q|  <request xmlns="">\n| +
          %Q|    <orderNumber>#{values[:order]}</orderNumber>\n| +
          %Q|    <cardholder>\n| +
          %Q|      <firstName>#{values[:holder_name]}</firstName>\n| +
          %Q|      <lastName>#{values[:holder_surname]}</lastName>\n| +
          %Q|      <street>#{values[:holder_street]}</street>\n| +
          %Q|      <zip>#{values[:holder_zip]}</zip>\n| +
          %Q|      <city>#{values[:holder_city]}</city>\n| +
          %Q|      <country>#{values[:holder_country]}</country>\n| +
          %Q|      <email>#{values[:holder_email]}</email>\n| +
          %Q|      <ip>#{values[:holder_ip]}</ip>\n| +
          %Q|    </cardholder>\n| +
          %Q|    <card>\n| +
          %Q|      <cardNumber>#{values[:card_number]}</cardNumber>\n| +
          %Q|      <expires>#{values[:card_exp]}</expires>\n| +
          %Q|      <cvv>#{values[:card_cvv]}</cvv>\n| +
          %Q|    </card>\n| +
          %Q|    <amount>\n| +
          %Q|      <value>#{values[:value]}</value>\n| +
          %Q|      <currency>#{values[:currency]}</currency>\n| +
          %Q|    </amount>\n| +
          %Q|    <product>\n| +
          %Q|      <productName>#{values[:product_name]}</productName>\n| +
          %Q|      <productUrl>#{values[:product_url]}</productUrl>\n| +
          %Q|    </product>\n| +
          %Q|  </request>\n| +
          %Q|</cpsxml>|
      end
  end
end

lib/complete_payment_systems/response_processing.rb
module CompletePaymentSystems
  class Response

    attr_reader :xml, :response_hash, :code, :message

    def initialize(xml)
      @xml = xml
      @response_hash = parse_response(@xml)
      @code = @response_hash["resultCode"]
      @message = @response_hash["resultMessage"]
    end

    def ok?
      return signature_ok? && message.match(CPS.config.success_regex).present? && code == "000"
    end

    def signature_ok?
      digi_signature = response_hash["digiSignature"]
      hash_string = build_hashable_string
      decoded_resp = Base64.decode64(digi_signature)
      public_key = OpenSSL::X509::Certificate.new(
        File.read "#{CPS.root}/lib/complete_payment_systems/certs/cps.cer").
        public_key
      return public_key.verify(OpenSSL::Digest::SHA1.new, decoded_resp, hash_string)
    end

    private

      def parse_response xml
        @response = Nokogiri::XML(xml) do |config|
          config.strict.nonet
        end

        return {
          "referenceId" => @response.xpath("//referenceId").text,
          "orderId" => @response.xpath("//orderId").text,
          "value" => @response.xpath("//value").text,
          "currency" => @response.xpath("//currency").text,
          "resultCode" => @response.xpath("//resultCode").text,
          "resultMessage" => @response.xpath("//resultMessage").text, # Important
          "resultText" => @response.xpath("//resultText").text, # Informative only
          "digiSignature" => @response.xpath("//digiSignature").text
        }
      end

      def build_hashable_string
        rh = @response_hash
        return "#{rh["referenceId"]}#{rh["orderId"]}#{rh["value"]}#{rh["currency"]}#{rh["resultCode"]}#{rh["resultMessage"]}#{rh["resultText"]}"
      end

  end

end

lib/complete_payment_systems.rb
require "complete_payment_systems/version"
require "complete_payment_systems/request_processing"
require "complete_payment_systems/response_processing"
#require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'base64'
require 'unirest'
require 'active_support/all'
require 'nokogiri'

module CompletePaymentSystems

  CPS = CompletePaymentSystems
  ROOT = File.expand_path("../..", __FILE__)

  def self.root
    ROOT
  end

  class << self
    attr_accessor :config
  end

  def self.configure
    self.config ||= Config.new
    yield(config)
  end

  class Config
    attr_accessor :default_user, :default_callback_url, :default_redirect_url, :default_product_name, :default_product_url, :default_country, :placeholder_value,
      :cps_url, :cps_method, :cert_pass, :rsa_cert_path, :success_regex

    def initialize
      @default_user = "pasta_test_3d"
      @default_callback_url = "http://www.google.com"
      @default_redirect_url = "http://www.google.lv"
      @default_product_name = "Product"
      @default_product_url = "www.test.com"
      @default_country = "US"
      @placeholder_value = "Not@Used.com"


      @cps_url = "https://3ds.cps.lv/GatorServo/request"
      @cps_method = "sendForAuth"
      @cert_pass = 'pasS%123'
      @rsa_cert_path = "#{CPS.root}/lib/complete_payment_systems/certs/Pasta_test_3d.pem"
      @success_regex = /Captured/
    end

  end

  CPS.configure {}

  def self.return_test_instance
    hash = {
      order: (Time.now.to_i),            # Pass the unique purchase ID here
      value: "166",                      # Pass the purchase value in cents here (1$ purcase value = 100)
      currency: "USD",    # Pass the purchase currency 3-letter code here ($ = "USD")
      holder_name: "John",               # Ask buyer for this in a form
      holder_surname: "Doe",             # Ask buyer for this in a form
      card_number: "4314229999999913",   # Ask buyer for this in a form
      card_exp: "01/18",                 # Ask buyer for this in a form
      card_cvv: "123",                   # Ask buyer for this in a form
      holder_ip: "123.124.125.226"       # Get this from request.remote_ip
    }

    return CPS::Request.new(hash)

  end

end

CPS = CompletePaymentSystems